devtools/x-core/src/git.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::errors::*;
+use determinator::Paths0;
+use guppy::{graph::PackageGraph, MetadataCommand};
+use log::info;
+use once_cell::sync::OnceCell;
+use std::{
+    borrow::Cow,
+    ffi::{OsStr, OsString},
+    fmt,
+    path::{Path, PathBuf},
+    process::{Command, Stdio},
+};
+
+/// Support for source control operations through running Git commands.
+///
+/// This assumes that the underlying Git repository doesn't change in the middle of an operation,
+/// and caches data as a result. If mutation operations are added, the caches would need to be
+/// invalidated.
+#[derive(Clone, Debug)]
+pub struct GitCli {
+    root: &'static Path,
+    // Caches.
+    tracked_files: OnceCell<Paths0>,
+}
+
+impl GitCli {
+    /// Creates a new instance of the Git CLI.
+    pub fn new(root: &'static Path) -> Self {
+        Self {
+            root,
+            tracked_files: OnceCell::new(),
```

Comment:
hmmm, if a dev grabs the source location (rather than a git pull, this wont detect the lack of a .git dir in the workspace and will fail later?   What would that error message look like?

Reply:
Chatting with Rex F2F, what he meant was the source archive -- definitely agreed that it needs to be handled.

Reply:
Added handling for missing .git.